Render <model> tags in Model process

eddydas opened the pull request:
#### 8a4b871eeecd5ca8e2001f76f417c739209a8c4f
<pre>
Render &lt;model&gt; tags in Model process
<a href="https://rdar.apple.com/124051126">rdar://124051126</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=270504">https://bugs.webkit.org/show_bug.cgi?id=270504</a>

Reviewed by Tim Horton.

Model process &lt;mdoel&gt; rendering path now uses the same WKSeparatedModelLayer
class as the in-process path to actually render the 3D model.

Also added required entitlement and sandbox rules to support rendering in
Model process.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:

Canonical link: <a href="https://commits.webkit.org/275725@main">https://commits.webkit.org/275725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/444f0f2efba5f7aed8cefa8cc06dbb5b955afb0e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42637 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/21659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45039 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45246 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38761 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/25321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19024 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35291 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43211 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/25321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36698 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16241 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/42508 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/25321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/694 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/25321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46741 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/17456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14382 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41999 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9528 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -84,9 +88,12 @@
-    RELEASE_LOG(ModelElement, "%p - ModelProcessModelPlayerProxy::loadModel size=%zu url=%s id=%" PRIu64, this, model->data()->size(), model->url().string().utf8().data(), m_id.toUInt64());
+    RELEASE_LOG(ModelElement, "%p - ModelProcessModelPlayerProxy::loadModel size=%zu id=%" PRIu64, this, model->data()->size(), m_id.toUInt64());
+    sizeDidChange(layoutSize);
+    [m_layer setPlayer:WebKit::SeparatedModelPlayer::create(true)];
```
Please adopt an enum class for this

hortont424 reviewed (approved):
r+ but the sandbox changes need Per Arne's review before landing

hortont424 commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
```diff
@@ -25,3 +25,27 @@
+(allow mach-lookup
+    (global-name "com.apple.surfboard.applicationservice"))
+(allow mach-lookup
+    (xpc-service-name "com.apple.ImageIOXPCService"))
```
More likely we should disable OOP ImageIO in this process like we do in the WP!

eddydas replied:
This line will be removed once the ImageIO work has been done.

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
```diff
@@ -25,3 +25,27 @@
+    (iokit-user-client-class "AGXDeviceUserClient"))
+
+(allow mach-lookup
+    (global-name "com.apple.audio.AudioSession"))
```
Is access to this audio service required?

eddydas replied:
Not for our current scope. I have removed this!

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
```diff
@@ -25,3 +25,27 @@
+    (global-name "com.apple.audio.AudioSession"))
+(allow mach-lookup
+    (global-name "com.apple.surfboard.applicationservice"))
+(allow mach-lookup
+    (xpc-service-name "com.apple.ImageIOXPCService"))
```
Perhaps we can add a FIXME to remove this when appropriate?

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
```diff
@@ -25,3 +25,27 @@
 
 (allow mach-lookup
     (global-name "com.apple.CARenderServer"))
+
+(allow iokit-open-user-client
+    (iokit-user-client-class "AppleParavirtDeviceUserClient"))
+(allow iokit-open-user-client
+    (iokit-user-client-class "AGXDeviceUserClient"))
```
Maybe we also can add a FIXME here to block these once that is possible?

eddydas committed:
Render <model> tags in Model process
rdar://124051126
https://bugs.webkit.org/show_bug.cgi?id=270504

Reviewed by Tim Horton.

Model process <mdoel> rendering path now uses the same WKSeparatedModelLayer
class as the in-process path to actually render the 3D model.

Also added required entitlement and sandbox rules to support rendering in
Model process.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:

Canonical link: https://commits.webkit.org/275725@main
